Add docs for useCallback (#586)
typescript-cheatsheets/react: README.md

This temporarily "lies" to the TypeScript compiler that `{}` is of type `User`. You should follow up by setting the `user` state — if you don't, the rest of your code may rely on the fact that `user` is of type `User` and that may lead to runtime errors.

#### useCallback

You can type the `useCallback` just like any other functions

```ts
const memoizedCallback = useCallback(
  (param1: string, param2: number) => {
    console.log(param1, param2)
    return { ok: true }
  },
  [...],
);
/**
 * VSCode will show the following type:
 * const memoizedCallback:
 *  (param1: string, param2: number) => { ok: boolean }
 */
```

> Note that for React < 18, the function signiture of `useCallback` is as follow, in which the args are typed as `any[]` as default.

Reviewer: ```suggestion
Note that for React < 18, the function signature of `useCallback` is as follow, in which the args are typed as `any[]` as default.
```

```ts
function useCallback<T extends (...args: any[]) => any>(
  callback: T,
  deps: DependencyList
): T;
```

> In React >= 18, the function signiture of `useCallback` changed to the following,

Reviewer: ```suggestion
In React >= 18, the function signature of `useCallback` changed to the following:
```

```ts
function useCallback<T extends Function>(callback: T, deps: DependencyList): T;
```

> Therefore the follow code will yield "`Parameter 'e' implicitly has an 'any' type.`" error in React >= 18, but not <17.

Reviewer: ```suggestion
Thereforee the following code will yield "`Parameter 'e' implicitly has an 'any' type.`" error in React >= 18, but not <17.
```

```ts

> useCallback((e) => {}, []);

Reviewer: ```suggestion
// @ts-expect-error Parameter 'e' implicitly has 'any' type.
useCallback((e) => {}, []);
// Explicit 'any' type.
useCallback((e: any) => {}, []);
```

```

> 
> See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210

Reviewer: ```suggestion
```

Not relevant since mostly technical

This temporarily "lies" to the TypeScript compiler that `{}` is of type `User`. You should follow up by setting the `user` state — if you don't, the rest of your code may rely on the fact that `user` is of type `User` and that may lead to runtime errors.

#### useCallback

> You can type the `useCallback` just like any other function.

Reviewer 2: ```suggestion
You can type `useCallback` just like any other function.
```

```ts
const memoizedCallback = useCallback(
  (param1: string, param2: number) => {
    console.log(param1, param2)
    return { ok: true }
  },
  [...],
);
/**
 * VSCode will show the following type:
 * const memoizedCallback:
 *  (param1: string, param2: number) => { ok: boolean }
 */
```

Note that for React < 18, the function signature of `useCallback` typed arguments as `any[]` by default:

```ts
function useCallback<T extends (...args: any[]) => any>(
  callback: T,
  deps: DependencyList
): T;
```

In React >= 18, the function signature of `useCallback` changed to the following:

```ts
function useCallback<T extends Function>(callback: T, deps: DependencyList): T;
```

> Therefore, the following code will yield "`Parameter 'e' implicitly has an 'any' type.`" error in React >= 18, but not <17.

Reviewer 2: ```suggestion
Therefore, the following code will yield "`Parameter 'e' implicitly has an 'any' type.`" error in React >= 18, but not in React < 18.
```
